### ISSUE TITLE
[Bug]: Input field becomes unresponsive after window.Alert() and window.confirm()

### ISSUE THREAD

**Issue by TC-kHa** [electron/electron#31917]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Electron Version

16.0.1

### What operating system are you using?

Windows

### Operating System Version

Windows 10

### What arch are you using?

x64

### Expected Behavior

The input field is clickable and responsive after the user makes a selection from a windows Alert or Confirm.

### Actual Behavior

The input field is unclickable. The workaround is to open/close devtools.

### Additional Information

It seems like this a on-going issue from other users.

https://github.com/electron/electron/issues/22923

https://stackoverflow.com/questions/56805920/cant-edit-input-text-field-after-window-alert/56806101#56806101

I'm able to reproduce this issue consistently from windows electron app

I was able to reproduce this issue from `electron-quick-start` repo with the latest electron build `16.0.1` using the following `index.html `code below

```
<!DOCTYPE html>
<!--Display Alert Message on Button Click Event.-->
<html>
    <head>
        <title>Display Alert Message on Button Click Event.</title>
        <script type="text/javascript">
            function showMessage() {
                alert("Hello friends, this is message.");
            }
        </script>
    </head>
    <body>
            <h1>Display Alert Message on Button Click Event.</h1>
            <b>Click on button to display message: </b><br />
            <br />
            <input type="button" id="btnShowMsg" value="Click Me!" onClick="showMessage()" />
            <input type="text"/>
    </body>
</html>
```

**Comment by ItsYanko** [comment]
same issue. electron 16.0.1, windows 11 64 bits.

switching to another app and then coming back to the electron app makes it responsive again, but it's only a temprary fix.

**Comment by Seeroy** [comment]
Same issue in 17.1.2

**Comment by lzx81647** [comment]
Same issue in 18.0.3

**Comment by OtavioSDev** [comment]
Same issue in 20.1.3

**Comment by keerthi16** [comment]
bump

**Comment by tjwrona** [comment]
still a problem in the version of electron I downloaded yesterday. (latest stable version)

**Comment by OStewart-dev** [comment]
The problem still exists (March 19, 2023) and I am also running the latest version of Electron.

**Comment by Mogeiman** [comment]
bump

**Comment by jacksonpm** [comment]
bump!

**Comment by alljinx** [comment]
Same issue with ^23.1.3

**Comment by Nick-Dunne** [comment]
Same issue with ^24.1.3

**Comment by mErlin-sp** [comment]
Same issue with ^25.1.1, win11 x64

**Comment by Hamza-Bandukwala** [comment]
Bump

**Comment by plauwaer** [comment]
Same issue with 25.2.0, win11 x64

**Comment by alljinx** [comment]
The workaround of @existical (thanks to you) works well. Here's the way I use it to make it esay to remove when the official fix be released : 

_Main process (same as @existical) :_ 
```
ipcMain.on('focus-fix', () => {
    win.blur();
    win.focus();
});
``` 

_Renderer (browser) process :_ 
```
const oldWindowConfirmFunc = confirm;
window.confirm = (message?: string) => {
  const res = oldWindowConfirmFunc(message);
  (window as unknown as { electron: { send: (msg: string) => void } }).electron.send('focus-fix');
  return res;
};
const oldWindowAlertFunc = alert;
window.alert = (message?: string) => {
  const res = oldWindowAlertFunc(message);
  (window as unknown as { electron: { send: (msg: string) => void } }).electron.send('focus-fix');
  return res;
};
```

Then you just use window.alert and window.confirm as it's not bugged at all and simply remove theses blocs when it is officially fixed.

**Comment by plauwaer** [comment]
> The workaround of @existical (thanks to you) works well. Here's the way I use it to make it esay to remove when the official fix be released :
> 
> _Main process (same as @existical) :_
> 
> ```
> ipcMain.on('focus-fix', () => {
>     win.blur();
>     win.focus();
> });
> ```
> 
> _Renderer (browser) process :_
> 
> ```
> const oldWindowConfirmFunc = confirm;
> window.confirm = (message?: string) => {
>   const res = oldWindowConfirmFunc(message);
>   (window as unknown as { electron: { send: (msg: string) => void } }).electron.send('focus-fix');
>   return res;
> };
> const oldWindowAlertFunc = alert;
> window.alert = (message?: string) => {
>   const res = oldWindowAlertFunc(message);
>   (window as unknown as { electron: { send: (msg: string) => void } }).electron.send('focus-fix');
>   return res;
> };
> ```
> 
> Then you just use window.alert and window.confirm as it's not bugged at all and simply remove theses blocs when it is officially fixed.

@alljinx: Thank you  the workaround of @existical indeed works well.

**Comment by HariAcharya80** [comment]
Bump

**Comment by pvrobays** [comment]
Still an issue, even with the latest electron version.
I've updated the gist I've created for #35129 which uses the `unload` event:
https://gist.github.com/37749f99b5872f49e99a9667a7a0e4f3

**Comment by nobertomaciel** [comment]
This issue yet persists on latest version 27.0.4 windows 10.

**Comment by nobertomaciel** [comment]
> stale

bump

**Comment by armysarge** [comment]
bump

**Comment by qkrwpdlr** [comment]
bump

```
window.alert = ()=>{}
```
: )

**Comment by AlyssaKit** [comment]
bump

**Comment by LsYueh** [comment]
Same issue with 29.1.0.
But now I'm using IPC to invoke `dialog.showMessageBox()`, instead of call `window.alert()` directly.

**Comment by mayur3896** [comment]
Bump

**Comment by khashayaramini** [comment]
Same issue with 29.1.4

**Comment by SakethSripada** [comment]
same issue in 31.0.1

**Comment by agustaf9** [comment]
Same issue with 31.3.1

**Comment by PedroPedalos** [comment]
bump, but why is there no reaction on this issue from 2021 yet?
It is quite a serious a problem!

**Comment by eyobed-dev** [comment]
> bump, but why is there no reaction on this issue from 2021 yet? It is quite a serious a problem!

Electron team decided to not support these as they are not asynchronous.

**Comment by fabiofabbri84** [comment]
If you are using Electron with Cordova, we have made a Cordova plugin that fix this issue: https://github.com/cimatti/cordova-electron-dialogs-fix

Basically it uses the same solution found by @existical and @plauwaer

**Comment by LqdBcnAtWork** [comment]
@eyobed-dev 
> > bump, but why is there no reaction on this issue from 2021 yet? It is quite a serious a problem!
> 
> Electron team decided to not support these as they are not asynchronous.

Could you share a link to this?

Not that I don't believe you, having a link to their statement would be nice.

Also, _I've written synchronous replacements to `alert()` and `confirm()` using their own architecture that doesn't have this problem._ I used the sync dialog methods. Zero problems, just a bit of boiler plate to pass the sync request to Main and the response back up.

It's honestly a little insulting that they can't spend the probably 10 minutes to fix this. I would, but I'm unfamiliar with C++ and the electron.js codebase. Especially if all @fabiofabbri84 needed to do is call a `blur()` and `focus()` on the `BrowserWindow` after the function call to fix it for Cordova apps.

**Comment by fabiofabbri84** [comment]
I found also this: #19977

**Comment by towd9** [comment]
bump, how come no one has found a fix to this? Mine happens whenever I delete a table row and try adding another one in or editing another existing one. The problem looks simple but i'm struggling to find some sort of workaround on my own

**Comment by ZarooriReturn** [comment]
same issue

**Comment by LqdBcnAtWork** [comment]
> `Confirm()` and `alert()` have been broken for a long time.
> Solution:
> Use electrons [dialog.showMessageBox()](https://www.electronjs.org/docs/latest/api/dialog#dialogshowmessageboxwindow-options) instead

I would hardly consider this a solution. A work-around? Definitely. True solution? No. 
I've build around this issue using Electron's dialog a few times now. But I'd still much rather have `alert`, `confirm` and `prompt` work as they're supposed to.

The existence of a work around does not mean a solution isn't needed. While I understand this issue being low priority. It still should be fixed.

I would do it myself, but it would take me more time than I have free to learn the codebase enough to even consider changing anything. Especially since someone who knows the codebase should be able to fix it in a day's work. Even if the 'fix' is causing a `blur` and `focus` event after the popup closes.

**Comment by BOBO-D24** [comment]
Bump

**Comment by paradite** [comment]
When using context bridge, here is a modified version of walkaround based on @alljinx and @existical's version:

```js
// mainWindow.ts
ipcMain.on('focus-fix', () => {
  window?.blur();
  window?.focus();
});

// preload/index.ts

contextBridge.exposeInMainWorld('electronAPI', {
  focusFix: () => ipcRenderer.send('focus-fix'),
});

// render.d.ts

export interface ElectronAPI {
  focusFix: () => void;
}

// renderer/src/index.ts

const oldWindowConfirmFunc = confirm;
window.confirm = (message?: string) => {
  const res = oldWindowConfirmFunc(message);
  console.log('focusFix');
  window.electronAPI.focusFix();
  return res;
};
const oldWindowAlertFunc = alert;
window.alert = (message?: string) => {
  const res = oldWindowAlertFunc(message);
  console.log('focusFix');
  window.electronAPI.focusFix();
  return res;
};

```

**Comment by ZarooriReturn** [comment]
Yes but I use the best way like I was facing this issue when we use the alert or confirmation dialogues then the input input field becomes disabled but when I use the custom alerts and confirmations like by creating separate due and separate look for alert and for confirmation then that's working fine so because there is no interaction with the default alert and confirm.

**Comment by KevinAyalaAguilera** [comment]
bump

**Comment by gduquesnay** [comment]
bump

**Comment by PLangari** [comment]
bump

**Comment by YoungjuneKwon** [comment]
Based on the implementations from [@paradite](https://github.com/paradite), [@alljinx](https://github.com/alljinx), and [@existical](https://github.com/existical), I resolved the issue by overriding the default alert and confirm function using the `contextBridge.executeInMainWorld` feature in `preload.js`.

```javascript
// main.ts
ipcMain.on('focus-fix', () => {
  win.blur();
  win.focus();
});

// preload.ts
const { contextBridge, ipcRenderer } = require('electron');

contextBridge.exposeInMainWorld('electronAPI', {
  focusFix: () => ipcRenderer.send('focus-fix'),
});

contextBridge.executeInMainWorld({func: () => {
  const oldWindowConfirmFunc = confirm;
  window.confirm = (message?: string) => {
    const res = oldWindowConfirmFunc(message);
    console.log('focusFix');
    window.electronAPI.focusFix();
    return res;
  };
  
  const oldWindowAlertFunc = alert;
  window.alert = (message?: string) => {
    const res = oldWindowAlertFunc(message);
    console.log('focusFix');
    window.electronAPI.focusFix();
    return res;
  };
}});
```

**Comment by z3niths** [comment]
Still experience the problem
electron : v36.2.1

**Comment by arcanisgk** [comment]
Bump, same problem when I delete a data entry in a table and need to reuse the form (Bootstrap modal) to continue adding records to the table, locked fields.

I already suggested we move to another technology. Electron is dead...

**Comment by paras0x** [comment]
Still a bug.

**Comment by nik736** [comment]
This is still a bug

**Comment by ghost** [comment]
Yes still a bug the simple solution is that using custom HTML,CSS,JS you create a custom alert or confirmation model never use browser default "alert or confirm" because in your project could be a site or anything else anywhere you used this then you need to close your app and reopen and ofcorce this is a poor way. So the best thing is remove all default alert and confirm etc and use custom. I am using custom it's great but yeah it's true if you're project is big then ofcorce this will take more time.

**Comment by wnunezc** [comment]
> Yes still a bug the simple solution is that using custom HTML,CSS,JS you create a custom alert or confirmation model never use browser default ...

The purpose of using Electron in desktop applications is to be able to take advantage of native features. If I wanted to program a 100% HTML+JS+CSS site, I would use a URL and a site and eliminate all dependency of Electron.

**Comment by ghost** [comment]
> > Yes still a bug the simple solution is that using custom HTML,CSS,JS you create a custom alert or confirmation model never use browser default ...
> 
> The purpose of using Electron in desktop applications is to be able to take advantage of native features. If I wanted to program a 100% HTML+JS+CSS site, I would use a URL and a site and eliminate all dependency of Electron.


I understand your point — Electron is meant to bring native-like capabilities to desktop apps. My concern here is not about abandoning Electron's native features altogether, but about avoiding certain problematic browser-native APIs (like the default alert() / confirm() dialogs) that are known to cause input locking issues in Electron’s rendering process.

Replacing them with custom modals doesn’t mean turning the whole app into a plain web site; it’s just a targeted workaround for a bug that has existed for years and still breaks UX in critical flows.

In larger apps, stability and predictable behavior often outweigh the convenience of native dialogs, especially when they block further user interaction until a full reload. So, this isn’t about removing Electron’s benefits — it’s about ensuring the parts we rely on work consistently in production.

**Comment by wnunezc** [comment]
> I understand your point — Electron is meant to bring native-like capabilities to desktop apps. My concern here is not about abandoning Electron's native features altogether, but about avoiding certain ...

@ghostofmars  What you're suggesting doesn't make sense. It's assumed that if a software has capabilities, those capabilities should at least not have persistent bugs over time. Indicating that third-party or alternative code must be implemented is incorrect; it goes in the other direction and doesn't justify abandoning support for the application, whether open source or paid.

What you're suggesting could become bad practice in the long run...

**Comment by ghost** [comment]
> > I understand your point — Electron is meant to bring native-like capabilities to desktop apps. My concern here is not about abandoning Electron's native features altogether, but about avoiding certain ...
> 
> @ghostofmars  What you're suggesting doesn't make sense. It's assumed that if a software has capabilities, those capabilities should at least not have persistent bugs over time. Indicating that third-party or alternative code must be implemented is incorrect; it goes in the other direction and doesn't justify abandoning support for the application, whether open source or paid.
> 
> What you're suggesting could become bad practice in the long run...
> 
> 

Yes, this is a long-standing bug that has been around for years without a proper fix. Ideally, it should be solved at the framework level — I completely agree on that. But when it isn’t being addressed for such a long time, the practical approach is to implement a stable workaround so the app’s UX isn’t blocked. This isn’t about abandoning Electron’s native capabilities, it’s about ensuring users can keep working without having to restart the app every time this bug is triggered. Hopefully you caught what I wanna say.

**Comment by SmatAmin** [comment]
> bump, how come no one has found a fix to this? Mine happens whenever I delete a table row and try adding another one in or editing another existing one. The problem looks simple but i'm struggling to find some sort of workaround on my own. Also, does de issue go away once the app is packaged/published?
> 
It doesn't.

**Comment by Jjarrard** [comment]
I'm using 38.1.2, having the same issue with alerts losing focus and unable to focus on a text element until I lose focus of the application completely and regain it. My solution (because it works for me) was to replace all "save success" alerts with DOM toasts

**Comment by JerryI** [comment]
Bump. Confirmed on 39.0.0 (Windows, MacOS rarely)

**Comment by fvspneto** [comment]
bump... confirmed  on 39.0.0 (Windows)

**Comment by JerryI** [comment]
I know this is not a solution, but I found a way to prevent this by using `dialog` API, which is safe:

```js
window.alert = electronAPI.openMessageWindow;
```

then after hooking `openMessageWindow` from the bridge to main process, you can call:

```js
function openMessageWindow(event, message) {
  dialog.showMessageBox({message: String(message)});
}
```

Then same can be done with `confirm` using promises, then it also won't block Javascript on the active window.

**Comment by JerryI** [comment]
> My solution (because it works for me) was to replace all "save success" alerts with DOM toasts

If you are using Electron anyway, I believe you should try native dialog windows, native drag events, keybindings as many API as the framework provides as possible. It can only improve UX.

**Comment by Jjarrard** [comment]
> > My solution (because it works for me) was to replace all "save success" alerts with DOM toasts
> 
> If you are using Electron anyway, I believe you should try native dialog windows, native drag events, keybindings as many API as the framework provides as possible. It can only improve UX.

The native dialogue window causes this bug.

**Comment by JerryI** [comment]
> > > My solution (because it works for me) was to replace all "save success" alerts with DOM toasts
> > 
> > 
> > If you are using Electron anyway, I believe you should try native dialog windows, native drag events, keybindings as many API as the framework provides as possible. It can only improve UX.
> 
> The native dialogue window causes this bug.

Then these may be unrelated issues. `window.alert` uses different API and also blocks Javascript completely (it also requires a window, while `dialog` API does not need it). In my case all issues were solved by switching to electron API.

**Comment by LqdBcnAtWork** [comment]
To be fair, blocking isn't the issue. Electron `dialog` vs window popups seemingly are.

I've build a sync replacement for popups using `dialog` with success.

Preload:
```js
const electronAPI = {
  /**
   * @param {string} text
   */
  alert: (text) => {
    return ipcRenderer.sendSync("alert", text);
  },

  /**
   * @param {string} text
   * @returns {boolean}
   */
  confirm: (text) => {
    return ipcRenderer.sendSync("confirm", text);
  },
}
```

main:
```js
    ipcMain.on("alert", (event, text) => {
      event.returnValue = dialog.showMessageBoxSync({
        message: text,
        type: "warning",
        title: "Alert",
      });
    });

    ipcMain.on("confirm", (event, text) => {
      event.returnValue =
        0 ==
        dialog.showMessageBoxSync({
          message: text,
          type: "question",
          title: "Confirm",
          buttons: ["OK", "Cancel"],
          defaultId: 0,
          cancelId: 1,
        });
    });
```

And it works as a drop-in replacement for `window.alert` & `window.confirm`.

So we can have synchronous version without any issues. Specifically the versions attached to `window` are an issue.

**Comment by agoudbg** [comment]
Can't believe 5 years have been passed and this issue is still here.

**Comment by JerryI** [comment]
> Can't believe 5 years have been passed and this issue is still here.

It might be a dead-end similar to https://github.com/electron/electron/issues/472
These features are not recommended to use I guess in general.

**Comment by LqdBcnAtWork** [comment]
> @agoudbg 
> Can't believe 5 years have been passed and this issue is still here.

I hate to break it to you, but we've had open issues for this for closer to 7 years. #19977 

Every version of Electron since v5 has had this issue. Recently I tested the last version of v4, and the first nightly of v5, which is exactly where the issue starts to show. I have a feeling something changed on the chromium side of things, and it's never been addressed in Electron because it's a low priority. Or if you're feeling cynical, because it doesn't affect VSCode.

I'd dive into it, but I have neither the time, nor the experience with C++ to do it.

**Comment by JerryI** [comment]
> To be fair, blocking isn't the issue. Electron `dialog` vs window popups seemingly are.
> 
> I've build a sync replacement for popups using `dialog` with success.
> 
> Preload:
> ```js
> const electronAPI = {
>   /**
>    * @param {string} text
>    */
>   alert: (text) => {
>     return ipcRenderer.sendSync("alert", text);
>   },
> 
>   /**
>    * @param {string} text
>    * @returns {boolean}
>    */
>   confirm: (text) => {
>     return ipcRenderer.sendSync("confirm", text);
>   },
> }
> ```
> 
> main:
> ```js
>     ipcMain.on("alert", (event, text) => {
>       event.returnValue = dialog.showMessageBoxSync({
>         message: text,
>         type: "warning",
>         title: "Alert",
>       });
>     });
> 
>     ipcMain.on("confirm", (event, text) => {
>       event.returnValue =
>         0 ==
>         dialog.showMessageBoxSync({
>           message: text,
>           type: "question",
>           title: "Confirm",
>           buttons: ["OK", "Cancel"],
>           defaultId: 0,
>           cancelId: 1,
>         });
>     });
> ```
> 
> And it works as a drop-in replacement for `window.alert` & `window.confirm`.
> 
> So we can have synchronous version without any issues. Specifically the versions attached to `window` are an issue.

This could be an ultimate solution, waiting for the fixes to `window` is pointless. As it was said implicitly, we should stick to APIs used in VSCode.

Please, pin that reply as an answer.

**Comment by leonardoarroyo** [comment]
At this point, given the age of this ticket and its numerous duplicates, it's fair to assume an upstream fix isn't coming anytime soon. If you're facing this, save yourself the headache: unless you really need the sync blocking behavior, I highly recommend bypassing native dialogs entirely and rolling out your own DOM-based alerts/prompts/confirms. If you need the synchronous behaviour, use the electron dialog API.

**Comment by JerryI** [comment]
👍 for native dialog async API
It is robust and more customisable

**Comment by LiquidBacn** [comment]
> At this point, given the age of this ticket and its numerous duplicates, it's fair to assume an upstream fix isn't coming anytime soon. If you're facing this, save yourself the headache: unless you really need the sync blocking behavior, I highly recommend bypassing native dialogs entirely and rolling out your own DOM-based alerts/prompts/confirms. If you need the synchronous behaviour, use the electron dialog API.

You actually don't need to give up on the sync nature of the built-ins. With sendSync ipc and the sync methods for dialogue boxes you can create a near drop in replacement.

At least, while the sync options continue to be supported...
